onnxruntime/core/providers/cpu/llm/attention.cc

```diff
@@ -0,0 +1,749 @@
+                                past_value_data,
+                                present_data,
+                                present_value_data,
+                                parameters.transpose_output,
+                                tp,
+                                past_present_share_buffer,
+                                max_sequence_length);
+
+  return Status::OK();
+}
+
+template <typename T>
+Tensor* AttentionBase<T>::GetPresent(OpKernelContext* context,
+                                     const Tensor* past,
+                                     int batch_size,
+                                     int head_size,
+                                     int kv_num_heads,
+                                     int kv_sequence_length,
+                                     int past_sequence_length) const {
+  // Input and output shapes:
+  //   past        : (2, batch_size, kv_num_heads, past_sequence_length, head_size)
+  //   present     : (2, batch_size, kv_num_heads, past_sequence_length + kv_sequence_length, head_size)
+
+  std::array<int64_t, 5> present_dims{2, batch_size, kv_num_heads, static_cast<int64_t>(kv_sequence_length) + past_sequence_length, head_size};
+
+  TensorShape present_shape(present_dims);
+  Tensor* present = context->Output(1, present_shape);
+  if (nullptr != past && nullptr == present) {
+    ORT_THROW("Expect to have present state output when past state input is given");
+  }
+
+  return present;
+}
+
+template <typename T, typename U>
+void make_copy(T* mask_data, const U* mask_index, size_t size);
+
+template <>
+void make_copy<float, float>(float* mask_data, const float* mask_index, size_t size) {
+  memcpy(mask_data, mask_index, size * sizeof(float));
+}
+
+template <>
+void make_copy<float, bool>(float* mask_data, const bool* mask_index, size_t size) {
+  for (size_t i = 0; i < size; ++i) {
+    mask_data[i] = mask_index[i] ? 0.0f : std::numeric_limits<float>::lowest();
+  }
+}
+
+template <typename T>
+template <typename U>
+void AttentionBase<T>::PrepareMask(const U* mask_index,
+                                   gsl::span<const int64_t> mask_index_dims,
+                                   T* mask_data,
+                                   bool causal,
+                                   int batch_size,
+                                   int sequence_length,
+                                   int kv_sequence_length,
+                                   int past_sequence_length) const {
+  const int all_sequence_length = past_sequence_length + kv_sequence_length;
+
+  // mask_data has been filled with 0, and its shape is BxSxT
+  T* p_mask = mask_data;
+
+  // 4D mask in Megatron GPT2 is currently not support in CPU kernel
+  if (nullptr != mask_index && mask_index_dims.size() == 4) {
+    ORT_NOT_IMPLEMENTED("4D mask in attention cpu kernel is not supported");
+  }
+
+  // For 3D mask, convert values 0 to mask_filter_value, and 1 to 0.0, then apply unidirectional mask if any.
```

Comment:
This code is for the contrib op. 
I think the Attention in onnx domain only support 4D mask.

Reply:
The documentation says:

**Attention mask**. Shape must be broadcastable to 4D tensor with shape (batch_size, q_num_heads, q_sequence_length, total_sequence_length) where total_sequence_length = past_sequence_length + kv_sequence_length. Two types of masks are supported. A boolean mask where a value of True indicates that the element should take part in attention. Also supports a float mask of the same type as query, key, value that is added to the attention score.